docs/api_overview.md

```diff
@@ -0,0 +1,472 @@
+out-of-band on what hostcalls/callbacks will be available. For API details, see
+associated doc comments in [proxy_wasm_externs.h] and [proxy_wasm_api.h].
+
+## Example
+
+Some of the concepts and APIs described above are illustrated by the example
+plugin code in [http_wasm_example.cc].
+
+First, the example plugin defines subclasses of `RootContext` and `Context`
+that override callbacks of interest:
+
+```c++
+class ExampleRootContext : public RootContext {
+public:
+  explicit ExampleRootContext(uint32_t id, std::string_view root_id) : RootContext(id, root_id) {}
+  bool onStart(size_t) override;
+  bool onConfigure(size_t) override;
+  // ...
+};
+
+class ExampleContext : public Context {
+public:
+  explicit ExampleContext(uint32_t id, RootContext *root) : Context(id, root) {}
+  void onCreate() override;
+  FilterHeadersStatus onRequestHeaders(uint32_t headers, bool end_of_stream) override;
+  // ...
+};
+```
+
+`RegisterContextFactory` registers these classes for creation:
+
+```c++
+static RegisterContextFactory register_ExampleContext(CONTEXT_FACTORY(ExampleContext),
+                                                ROOT_FACTORY(ExampleRootContext),
+                                                "my_root_id");
+```
+
+The implementations of callback methods use hostcalls to access and
+mutate request or response state:
+
+```c++
+FilterHeadersStatus ExampleContext::onResponseHeaders(uint32_t, bool) {
+  LOG_DEBUG(std::string("onResponseHeaders ") + std::to_string(id()));
+  auto result = getResponseHeaderPairs();
+  auto pairs = result->pairs();
+  LOG_INFO(std::string("headers: ") + std::to_string(pairs.size()));
+  for (auto &p : pairs) {
+    LOG_INFO(std::string(p.first) + std::string(" -> ") + std::string(p.second));
+  }
+  addResponseHeader("X-Wasm-custom", "FOO");
+  replaceResponseHeader("content-type", "text/plain; charset=utf-8");
+  removeResponseHeader("content-length");
+  return FilterHeadersStatus::Continue;
+}
+```
+
+## Codemap
+
+The main files containing the Proxy-Wasm C++ SDK API and implementation are
+listed below:
+
+* [proxy_wasm_api.h]: main SDK API definition and implementation
+* [proxy_wasm_externs.h]: declarations for ABI-level hostcalls and callbacks
+* [proxy_wasm_common.h](../proxy_wasm_common.h): supporting types for the API
+* [proxy_wasm_enums.h](../proxy_wasm_enums.h): supporting enums for the API
+* [proxy_wasm_intrinsics.js](../proxy_wasm_intrinsics.js): list of Proxy-Wasm
+  ABI hostcalls, for use by [Emscripten](https://emscripten.org)
+* [proxy_wasm_intrinsics.proto](../proxy_wasm_intrinsics.proto): protobuf types
+  needed for gRPC calls
+* [proxy_wasm_intrinsics.h](../proxy_wasm_intrinsics.h): combined header file
+  that includes all other header files
+* [proxy_wasm_intrinsics.cc](../proxy_wasm_intrinsics.cc): implementation of
+  dispatch from ABI-level Proxy-Wasm callbacks to [Context] and [RootContext]
+  callback methods.
```

Comment:
Nit: Maybe move all link targets to the bottom of this file for consistency?

Reply:
Done